master/buildbot/locks.py

```diff
@@ -96,17 +96,19 @@ def isAvailable(self, requester, access):
 
         if access.mode == 'counting':
             # Wants counting access
-            return num_excl == 0 and num_counting + len(ahead) < self.maxCount \
+            return num_excl <= 1 - access.count \
```

Comment:
Here we should count the total of the lock counts in the queue ahead of us. We'll also need to have a unit test to ensure it works correctly.

Reply:
I hope it's ok like the latest change (apart from tests).